File: backend/app/services/wiring.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.

"""Resolve sibling references between a fleet's resources.

Components, transform stages, and routes are wired to each other by **id lists in
JSON columns** (``inputs_json``, ``source_ids_json``, branch ``sink_ids``,
``passthrough_sink_ids_json``), not foreign keys — so the database can't stop you
from deleting something another resource still points at. :func:`find_references`
computes that graph so delete endpoints can refuse (or warn) instead of silently
leaving a dangling reference that the renderer drops on the next deploy.
"""

from fastapi import HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.component import Component
from app.models.route import Route
from app.models.transform_stage import TransformStage
from app.services.config_render import _json_list


async def find_references(
    fleet_id: str, target_id: str, db: AsyncSession
) -> list[dict[str, str]]:
    """Return the fleet resources whose wiring references ``target_id`` — i.e. what
    would dangle if it were deleted. Each entry is ``{kind, id, name}``.

    Matches an exact id, or (for a route target) an input of the form
    ``<target_id>.<branch>`` so a route's branch outputs count as references too.
    """
    refs: list[dict[str, str]] = []

    def _matches(ids: list[object]) -> bool:
        for i in ids:
            if i == target_id or (isinstance(i, str) and i.startswith(f"{target_id}.")):
                return True
        return False

    comps = (
        (await db.execute(select(Component).where(Component.fleet_id == fleet_id)))
        .scalars()
        .all()
    )
    for c in comps:
        if c.id != target_id and _matches(_json_list(c.inputs_json)):
            refs.append({"kind": c.kind, "id": c.id, "name": c.name})

    stages = (
        (
            await db.execute(
                select(TransformStage).where(TransformStage.fleet_id == fleet_id)
            )
        )
        .scalars()
        .all()
    )
    for s in stages:
        if s.id != target_id and _matches(_json_list(s.inputs_json)):
            refs.append({"kind": "transform", "id": s.id, "name": s.name})

    routes = (
        (await db.execute(select(Route).where(Route.fleet_id == fleet_id)))
        .scalars()
        .all()
    )
    for r in routes:
        if r.id == target_id:
            continue
        ids: list[object] = list(_json_list(r.source_ids_json))
        ids += list(_json_list(r.passthrough_sink_ids_json))
        for b in _json_list(r.branches_json):
            if isinstance(b, dict):
                ids += list(b.get("sink_ids", []) or [])
        if _matches(ids):
            refs.append({"kind": "route", "id": r.id, "name": r.name})

    return refs


async def ensure_deletable(
    fleet_id: str | None, target_id: str, force: bool, db: AsyncSession
) -> None:
    """Raise 409 if ``target_id`` is still referenced, unless ``force`` is set.

    The 409 ``detail`` carries the referencing resources so the UI can list them
    ("In use by Route 'fanout', Stage 'enrich'") and offer a force-delete.
    """
    if force or fleet_id is None:
        return
    refs = await find_references(fleet_id, target_id, db)
    if refs:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail={
                "message": "In use by "
                + ", ".join(f"{r['kind']} '{r['name']}'" for r in refs)
                + ". Unwire it first, or force-delete.",
                "references": refs,
            },
        )

File: frontend/src/lib/wiring.ts
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

/**
 * Client-side wiring graph helpers — the mirror of the backend `find_references`,
 * computed from the fleet data the Flow canvas already loads. Used by the node
 * drawer to show "Fed by" (a node's inputs) and "Feeds" (its consumers).
 */

import type { Component, Route, TransformStage } from '@/lib/types'

export interface FleetGraph {
  components: Component[]
  routes: Route[]
  stages: TransformStage[]
}

export type NodeKind = 'source' | 'sink' | 'transform' | 'route'
export interface NodeRef {
  kind: NodeKind
  id: string
  name: string
}

function matchesTarget(ids: string[] | undefined, targetId: string): boolean {
  return (ids ?? []).some((i) => i === targetId || i.startsWith(`${targetId}.`))
}

/** Resolve an id (component / stage / route, or a `routeId.branch` output) to a NodeRef. */
export function resolve(g: FleetGraph, id: string): NodeRef | null {
  const c = g.components.find((x) => x.id === id)
  if (c) return { kind: c.kind, id: c.id, name: c.name }
  const s = g.stages.find((x) => x.id === id)
  if (s) return { kind: 'transform', id: s.id, name: s.name }
  const r = g.routes.find((x) => x.id === id)
  if (r) return { kind: 'route', id: r.id, name: r.name }
  if (id.includes('.')) {
    const base = id.split('.')[0]
    const rr = g.routes.find((x) => x.id === base)
    if (rr) return { kind: 'route', id: rr.id, name: `${rr.name} → ${id.split('.').slice(1).join('.')}` }
  }
  return null
}

/** Nodes that read this node's output (consumers) — reverse of the wiring. */
export function feeds(g: FleetGraph, targetId: string): NodeRef[] {
  const out: NodeRef[] = []
  for (const c of g.components)
    if (c.id !== targetId && matchesTarget(c.inputs, targetId))
      out.push({ kind: c.kind, id: c.id, name: c.name })
  for (const s of g.stages)
    if (s.id !== targetId && matchesTarget(s.inputs, targetId))
      out.push({ kind: 'transform', id: s.id, name: s.name })
  for (const r of g.routes) {
    if (r.id === targetId) continue
    const ids = [
      ...(r.source_ids ?? []),
      ...(r.passthrough_sink_ids ?? []),
      ...(r.branches ?? []).flatMap((b) => b.sink_ids ?? []),
    ]
    if (matchesTarget(ids, targetId)) out.push({ kind: 'route', id: r.id, name: r.name })
  }
  return out
}

/** What feeds this node (its declared inputs), resolved to names. */
export function fedBy(g: FleetGraph, targetId: string): NodeRef[] {
  const inputIds =
    g.components.find((c) => c.id === targetId)?.inputs ??
    g.stages.find((s) => s.id === targetId)?.inputs ??
    g.routes.find((r) => r.id === targetId)?.source_ids ??
    []
  return inputIds.map((id) => resolve(g, id) ?? { kind: 'source', id, name: id })
}
